src/include/migraphx/generic_float.hpp

```diff
@@ -0,0 +1,392 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <migraphx/config.hpp>
+#include <migraphx/bit_cast.hpp>
+#include <algorithm>
+#include <limits>
+#include <iostream>
+#include <bitset>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+template<unsigned int N>
+constexpr unsigned int all_ones() noexcept
+{
+    return (1 << N) - 1;
+}
+
+struct float32_parts 
+{
+    unsigned int mantissa : 23;
+    unsigned int exponent : 8;
+    unsigned int sign : 1;
+
+    static constexpr unsigned int mantissa_width()
+    {
+        return 23;
+    }
+
+    static constexpr unsigned int max_exponent()
+    {
+        return all_ones<8>();
+    }
+
+    static constexpr int exponent_bias()
+    {
+        return all_ones<7>();
+    }
+
+    constexpr float to_float() const noexcept
+    {
+        return migraphx::bit_cast<float>(*this);
+    }
+};
+
+constexpr float32_parts get_parts(float f)
+{
+    return migraphx::bit_cast<float32_parts>(f);
+}
+
+
+template<unsigned int MantissaSize, unsigned int ExponentSize, unsigned int Flags = 0>
+struct __attribute__((packed)) generic_float
+{
+    unsigned int mantissa : MantissaSize;
+    unsigned int exponent : ExponentSize;
+    unsigned int sign : 1;
+
+    static constexpr int exponent_bias()
+    {
+        return all_ones<ExponentSize - 1>();
+    }
+
+    explicit constexpr generic_float(float f = 0.0) noexcept
+    {
+        from_float(get_parts(f));
+    }
+
+    constexpr generic_float &operator=(float f) noexcept
+    {
+        from_float(get_parts(f));
+        return *this;
+    }
+
+    constexpr generic_float operator-() const noexcept
+    {
+        generic_float result = *this;
+        result.sign = !this->sign;
+        return result;
+    }
+
+    constexpr float to_float() const noexcept
+    {
+        float32_parts f{};
+
+        f.sign = sign;
+
+        if(exponent == 0)
+        {
+
+            if(mantissa == 0)
+            {
+
+                f.exponent = 0;
+                f.mantissa = 0;
+            }
+            else
+            {
+                int shift  = 0;
+                f.mantissa = mantissa;
+
+                while((f.mantissa & (1 << MantissaSize)) == 0)
+                {
+                    f.mantissa <<= 1;
+                    shift++;
+                }
```

Comment:
This looks like its trying to count the bits. I think it would be a good idea to add a `popcount` function to do this(C++20 provides one but we can't use it yet) using Brian Kernighan's algorithm:

```cpp
constexpr unsigned int popcount(unsigned int v)
{
    unsigned int c = 0;
    for (; v; c++)
        v &= v - 1;
    return c;
}
```

Then you could just do:

```cpp
auto shift = popcount(mantissa);
f.mantissa = mantissa << shift;
```

Reply:
Working on integrating this

Reply:
Actually, looking at this again, I dont know if this was correct. It looks more like you are trying to count the bits after `MantissaSize` until it reaches a bit that is 0. It does seems a little strange because I would assume you would want to check the highest bit that is set(which it doesnt like thats what this loops does). Perhaps we need the [`countl_zero`](https://en.cppreference.com/w/cpp/numeric/countl_zero) function instead(but we will need to write our own implementation).

Reply:
I think with std::countl_zero would assume we're counting the amount of 0's from the end of f.mantissa size (which is 23 believe), not MantissaSize (which is 10)

Reply:
So we would need to pass in MantissaSize to this function. Is that ok?

Reply:
> I think with std::countl_zero would assume we're counting the amount of 0's from the end of f.mantissa size (which is 23 believe), not MantissaSize (which is 10)

First you can pass `mantissa` instead of `f.mantissa`(because it hasn't been assigned yet). Also, this will count the zeros for an `unsigned int` which is 32 bits. So you would need to subtract off the difference: `(32 - MantissaSize)`.

> So we would need to pass in MantissaSize to this function. Is that ok?

No the function should match the signature used for C++20 so we can replace it with that function in the future.